### PR TITLE
Update Go to 1.16

### PR DESCRIPTION
**Pull request by tklauser**
In preparation, clean up and rework some of the Makefile magic after #15023 was merged. In addition, make sure we use version 1.16 when referring to the upstream Go toolchain (i.e. the tarball downloaded from golang.org) but use 1.16.0 when referring to the Docker image.

See individual commits for details.

Second try for #15066 which I closed by accident.

### CODE REVIEW

**Review by christarazi — APPROVED** [review]
:tada:

**Review by sayboras — APPROVED** [review]

**Review by rolinh — APPROVED** [review]

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
retest-gke

**Comment by tklauser** [comment]
test-me-please

**Comment by tklauser** [comment]
retest-gke

**Comment by tklauser** [comment]
Hit #12756 on Travis CI: https://travis-ci.com/github/cilium/cilium/jobs/485470015

File: test/k8sT/manifests/test-verifier.yaml [hunk 1]
```diff
@@ -5,7 +5,7 @@ metadata:
 spec:
   containers:
   - name: cilium-builder
-    image: quay.io/cilium/cilium-runtime:4069625faec39c7dbae5b0b7c6271745a7304e35@sha256:77e44417ad7a56fb4cb00a10c35d240d455f14976b0f80b9ebafe83d050af571
+    image: quay.io/cilium/cilium-builder::6fe0cc8e74607c48e92758e1ff13bd267b70a29d@sha256:3a82c1bc13cf940524546d510643542d8cf63925ca9dc07ee8d2a63b7d14451
```
Comment (fristonio) [inline test/k8sT/manifests/test-verifier.yaml]:
```suggestion
    image: quay.io/cilium/cilium-builder:6fe0cc8e74607c48e92758e1ff13bd267b70a29d@sha256:3a82c1bc13cf940524546d510643542d8cf63925ca9dc07ee8d2a63b7d14451b
```

**Review by fristonio — CHANGES_REQUESTED** [review]
Looks good overall, seems to be a typo in the base image for `test-verifier` manifest. :rocket:

File: test/k8sT/manifests/test-verifier.yaml (hunk 1 above)
Comment (tklauser) [reply test/k8sT/manifests/test-verifier.yaml]: 🤦

Comment (tklauser) [reply test/k8sT/manifests/test-verifier.yaml]: Thanks @fristonio. Should have had my ☕ before pushing.

**Comment by tklauser** [comment]
test-me-please

**Review by fristonio — APPROVED** [review]
:shipit:

**Review by qmonnet — APPROVED** [review]
Looks all good to me, thanks for the doc update

**Review by aanm — APPROVED** [review]

**Pull request merged**